hudi-spark-datasource/hudi-spark3.2plus-common/src/main/scala/org/apache/spark/sql/hudi/source/SparkBatch.scala

```diff
@@ -0,0 +1,351 @@
+
+import java.net.URI
+
+class SparkBatch(spark: SparkSession,
+                 selectedPartitions: Seq[PartitionDirectory],
+                 partitionSchema: StructType,
+                 requiredSchema: StructType,
+                 filters: Seq[Filter],
+                 options: Map[String, String],
+                 @transient hadoopConf: Configuration) extends Batch with Serializable {
+
+  override def planInputPartitions(): Array[InputPartition] = {
+    // TODO support more accurate task planning.
+    val maxSplitBytes = FilePartition.maxSplitBytes(spark, selectedPartitions)
+    val splitFiles = selectedPartitions.flatMap { partition =>
+      partition.files.flatMap { file =>
+        val filePath = file.getPath
+        PartitionedFileUtil.splitFiles(
+          sparkSession = spark,
+          file = file,
+          filePath = filePath,
+          isSplitable = true,
+          maxSplitBytes = maxSplitBytes,
+          partitionValues = partition.values
+        )
+      }.toArray.sortBy(_.length)(implicitly[Ordering[Long]].reverse)
+    }
+    FilePartition.getFilePartitions(spark, splitFiles, maxSplitBytes).toArray
+
+  }
+
+  override def createReaderFactory(): PartitionReaderFactory = {
+    hadoopConf.set(ParquetInputFormat.READ_SUPPORT_CLASS, classOf[ParquetReadSupport].getName)
+    hadoopConf.set(
+      ParquetReadSupport.SPARK_ROW_REQUESTED_SCHEMA,
+      requiredSchema.json)
+    hadoopConf.set(
+      ParquetWriteSupport.SPARK_ROW_SCHEMA,
+      requiredSchema.json)
+    hadoopConf.set(
+      SQLConf.SESSION_LOCAL_TIMEZONE.key,
+      spark.sessionState.conf.sessionLocalTimeZone)
+    hadoopConf.setBoolean(
+      SQLConf.NESTED_SCHEMA_PRUNING_ENABLED.key,
+      spark.sessionState.conf.nestedSchemaPruningEnabled)
+    hadoopConf.setBoolean(
+      SQLConf.CASE_SENSITIVE.key,
+      spark.sessionState.conf.caseSensitiveAnalysis)
+
+    ParquetWriteSupport.setSchema(requiredSchema, hadoopConf)
+
+    // Sets flags for `ParquetToSparkSchemaConverter`
+    hadoopConf.setBoolean(
+      SQLConf.PARQUET_BINARY_AS_STRING.key,
+      spark.sessionState.conf.isParquetBinaryAsString)
+    hadoopConf.setBoolean(
+      SQLConf.PARQUET_INT96_AS_TIMESTAMP.key,
+      spark.sessionState.conf.isParquetINT96AsTimestamp)
+    val broadcastedConf = spark.sparkContext.broadcast(
+      new SerializableConfiguration(hadoopConf))
+    val sqlConf = spark.sessionState.conf
+    ReaderFactory(sqlConf, broadcastedConf)
+  }
+
+  case class ReaderFactory(sqlConf: SQLConf,
+                           broadcastedConf: Broadcast[SerializableConfiguration]) extends FilePartitionReaderFactory {
+    // TODO: if you move this into the closure it reverts to the default values.
+    // If true, enable using the custom RecordReader for parquet. This only works for
+    // a subset of the types (no complex types).
+    val resultSchema: StructType = StructType(partitionSchema.fields ++ requiredSchema.fields)
```

Comment:
Do we really need to clone whole Hudi's custom ParquetFileFormat impl here?

Reply:
There is some minor change to construct createReaderFactory, we may extend suitable action to `support more accurate task planning`

Reply:
Sorry, i still i'm not sure i understand what we're copying this for. Can you please elaborate what exactly you're trying to modify here?

We should copy code from Spark only in exceptional circumstances when there's just no other way around it. Otherwise we should avoid doing that at all costs.